### ISSUE TITLE
🐛 Bug: `unique-dependencies` doesn't flag duplicates between dependencies and devDependencies

### ISSUE THREAD

**Issue by sunnytsang1998** [JoshuaKGoldberg/eslint-plugin-package-json#1007]
### Bug Report Checklist

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have pulled the latest `main` branch of the repository.
- [x] I have [searched for related issues](https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/issues?q=is%3Aissue) and found none that matched my issue.

### Expected

I am specifying the same dependency in both `dependencies` and `devDependencies`, ie:
```json
{
  "name": "my-project",
  "dependencies": {
    "typescript": "^5.7.3",
  },
  "devDependencies": {
    "typescript": "^5.7.3",
  },
}
```

I would expect this to throw an error as we are using the same dependency in `dependencies` and `devDependencies`, where the dependency specified in `devDependencies` would be redundant.

### Actual

However, above `package.json` does not throw any error related to `package-json/unique-dependencies`. 

### Additional Info

I have tried to do some quick testing around, and managed to fix it quickly on my side, after slight modification on `src/rules/unique-dependencies.ts`.

Current code:
```ts
// Other code ...
export const rule = createRule({
  create(context) {
    function check(
      elements: (JsonAST.JSONNode | null)[],
      getNodeToRemove: (element: JsonAST.JSONNode) => JsonAST.JSONNode,
    ) {
      const seen = new Set();
      for (const element of elements
        .filter(isNotNullish)
// Other code ...
```

Modified code (I moved the `const seen = new Set();` line out of the `check` function block):
```ts
// Other code ...
export const rule = createRule({
  create(context) {
    const seen = new Set();
    function check(
      elements: (JsonAST.JSONNode | null)[],
      getNodeToRemove: (element: JsonAST.JSONNode) => JsonAST.JSONNode,
    ) {
      for (const element of elements
        .filter(isNotNullish)
// Other code ...
```

The problem seems to come from this line `const seen = new Set();`, which is being initialized inside the `check` function block, and as a result we failed to compare the list of dependencies included inside `dependencies` and `devDependencies`.

**Comment by michaelfaith** [comment]
Thanks for raising the issue.  Yeah, the rule was never intended to cross dependency type boundaries, because some pairings are actually acceptable, and in fact necessary.  For instance, packages that declare `peerDependencies` will almost always have them also declared in their `devDependencies` for local development.  That's a case that is extremely common and we shouldn't flag.  So, unfortunately a fix to address this wouldn't be as simple as just moving the `Set` to a higher scope.  

Even though the rule never intended to cover cross group duplicates, I do agree that having dupes across `dependencies` and `devDependencies` would be considered redundant.  So, happy to accept a PR, to solve that issue that doesn't also introduce other negative side effects.

**Comment by michaelfaith** [comment]
@all-contributors please add @sunnytsang1998 for bug

**Comment by michaelfaith** [comment]
Interestingly the docs to suggest that it should work across type boundaries, but there aren't any unit tests for that scenario, and the code clearly doesn't support it.  So, the documentation is certainly misleading in this case.

**Comment by michaelfaith** [comment]
The original issue other didn't end up having time to complete the work on this, if anyone else wants to take a stab at addressing it.

**Comment by Netail** [comment]
> The original issue author didn't end up having time to complete the work on this, if anyone else wants to take a stab at addressing it.

Sadly I don't use ESLint anymore, but am importing this rule (combined with eslint-plugin-package-json-dependencies' duplicate dependencies) to Biome, where the plan is to include it.

But it might be good to at least change the docs, in case someone else gets confused.
